rasa/architecture_prototype/graph.py

```diff
@@ -0,0 +1,309 @@
+        self,
+        component_class: Any,
+        config: Dict[Text, Any],
+        fn_name: Text,
+        node_name: Text,
+        inputs: Dict[Text, Text],
+        constructor_name: Text = None,
+        eager: bool = True,
+        persist: bool = True,
+        cache: Optional["TrainingCache"] = None,
+    ) -> None:
+        self._eager = eager
+        self._inputs = inputs
+        self._constructor_name = constructor_name
+        self._component_class = component_class
+        self._config = config
+        self._fn_name = fn_name
+        self._run_fn = getattr(self._component_class, fn_name)
+        self._component = None
+        self._node_name = node_name
+        self._persist = persist
+        self._cache = cache
+
+        if self._constructor_name:
+            self._constructor_fn = getattr(
+                self._component_class, self._constructor_name
+            )
+        else:
+            self._constructor_fn = self._component_class
+
+        input_names = list(inputs.keys())
+        self.validate_params_in_inputs(input_names, self._run_fn)
+        if not eager:
+            self.validate_params_in_inputs(input_names, self._constructor_fn)
+
+        if self._eager:
+            self.create_component(**self._config)
+
+    def validate_params_in_inputs(self, input_names, func):
+        params = inspect.signature(func).parameters
+        for param_name, param in params.items():
+            if param_name in ["self", "args", "kwargs", "persistor"]:
+                continue
+            if param.default is inspect._empty:
+                if param_name not in input_names:
+                    raise ValueError(
+                        f"{param_name} for function {func} is missing from inputs"
+                    )
+
+    def __call__(self, *args: Any) -> Dict[Text, Any]:
+        received_inputs = dict(ChainMap(*args))
+
+        # TODO: make the cache stuff a decorator
+        fingerprint_key = None
+        if self._cache:
+            fingerprint_key = self._cache.calculate_fingerprint_key(
+                # TODO: do list parts nicer
+                self._node_name,
+                self._config,
+                [list(v.values())[0] for v in args],
+            )
+
+        kwargs = {}
+
+        for input, input_node in self._inputs.items():
+            kwargs[input] = received_inputs[input_node]
+
+        if not self._eager:
+            const_kwargs = rasa.shared.utils.common.minimal_kwargs(
+                kwargs, self._constructor_fn
+            )
+            self.create_component(**const_kwargs, **self._config)
+
+        run_kwargs = kwargs
+
+        if "kwargs" not in rasa.shared.utils.common.arguments_of(self._run_fn):
+            run_kwargs = rasa.shared.utils.common.minimal_kwargs(kwargs, self._run_fn)
+
+        print(f"************** {self._node_name} ***************")
+        #  This alters the input
+        result = self._run_fn(self._component, **run_kwargs)
```

Comment:
I don't think it would be crazy to deepcopy all the inputs now. If you run it in multi-process mode it will be doing this anyway...

Reply:
I just realized that we probably broke predictions when doing deepcopying as we are relying on the changed tracker atm.